src/relay/op/tensor/reduce.cc

```diff
@@ -0,0 +1,209 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file reduce.cc
+ * \brief Reduction operators.
+ */
+#include <tvm/relay/expr.h>
+#include <tvm/relay/op.h>
+#include <numeric>
+#include "../type_relations.h"
+
+namespace tvm {
+namespace relay {
+
+/*! \brief Attributes for Reduce operators */
+struct ReduceAttrs : public tvm::AttrsNode<ReduceAttrs> {
+  Array<IndexExpr> axis;
+  bool keepdims;
+  bool exclude;
+
+  TVM_DECLARE_ATTRS(ReduceAttrs, "relay.attrs.ReduceAttrs") {
+    TVM_ATTR_FIELD(axis).set_default(Array<IndexExpr>({}))
+        .describe(R"code(The axis or axes along which to perform the reduction.
+
+      The default, `axis=()`, will compute over all elements into a
+      scalar array with shape `(1,)`.
+
+      If `axis` is int, a reduction is performed on a particular axis.
+
+      If `axis` is a tuple of ints, a reduction is performed on all the axes
+      specified in the tuple.
+
+      If `exclude` is true, reduction will be performed on the axes that are
+      NOT in axis instead.)code");
+
+    TVM_ATTR_FIELD(keepdims).set_default(false)
+      .describe("If this is set to `True`, the reduced axes are left "
+                "in the result as dimension with size one.");
+    TVM_ATTR_FIELD(exclude).set_default(false)
+      .describe("Whether to perform reduction on axis that are NOT in axis instead.");
+  }
+};
+
+/*!
+* \brief GetReduceAxes, get the new axis from indim and other arguments
+* \param indim Number of dimensions of input data.
+* \param axis The input axis vector.
+* \param exclude Whether 'axis' input given is the excluded axis.
+* \return r_axes The new reduced axes of the output.
+*/
+inline std::vector<int64_t> GetReduceAxes(const uint32_t indim,
+                                          const Array<IndexExpr>& inaxis,
+                                          bool exclude) {
+  if (!inaxis.defined()) {
+    std::vector<int64_t> r_axes(indim);
+    std::iota(r_axes.begin(), r_axes.end(), 0);
+    return r_axes;
+  }
+
+  std::vector<int64_t> in_axes;
+  for (auto i : inaxis) {
+    auto k = as_const_int(i);
+    CHECK(k != nullptr) << "Reduce axis need to be constant, cannot be symbolic";
+    auto axis = k[0];
```

Comment:
int64_t, while it is good to use auto, having more type information helps as long as it is not hard to type it out